### DISCUSSION TITLE
High dimensional neural network potentials

### DISCUSSION THREAD

**Discussion by cschran** [cp2k/cp2k#2595] [Team Posts]
Dear all,

@faolane and I have been working on a cp2k implementation of Behler-Parrinello style [neural network potentials](https://journals.aps.org/prl/abstract/10.1103/PhysRevLett.98.146401) that use [atom-centered symmetry functions](https://doi.org/10.1063/1.3553717). Up to now, our implementation is an in-house version of the Marx group. It supports the original RuNNer input format that is also used in [n2p2](https://compphysvienna.github.io/n2p2/), an open source fitting code. In addition, it makes use of most of the speed up ideas described in the recently published [NNP Lammps implementation](https://doi.org/10.1021/acs.jctc.8b00770). Thus, it scales quite well and gives convincing performance in the tests we have run so far.

We plan to merge our branch with the global cp2k repo, but would like to introduce our coding strategy first, before pushing too many changes at once:
1. The implementation relies on a new force environment called NNP.
2. The core routine calc_energy_force is parallelized over the atoms in the system, which have independent energy and force contributions in the NNP formalism.
3. The two main steps of this routine are the calculation of the atomic descriptors (atom-centered symmetry functions) and the actual evaluation of the artificial neural network.
4. We would like to offload most of the data structure and actual computations to a library in order to keep a tight integration between cp2k and an in-house fitting code. For that particular part, a cp2k sub-repo could be useful.

In particular 4. is still ongoing and we would appreciate any feedback on these plans.

All the best,

Christoph

**Comment by oschuett** [comment]
Hi Christoph,

it's exciting to see NNP arrive in CP2K!

Regarding 4., I'd advice you to put the code into the CP2K main repository where it will be [tested extensively](https://dashboard.cp2k.org). Furthermore, your feature will always be available since there is no need to install a separate library. 

That being said, if you really prefer a separate repro we can do that too.

-Ole

**Comment by dev-zero** [comment]
>     We would like to offload most of the data structure and actual computations to a library in order to keep a tight integration between cp2k and an in-house fitting code. For that particular part, a cp2k sub-repo could be useful.
> 
> In particular 4. is still ongoing and we would appreciate any feedback on these plans.

I guess this would be a similar approach like we have currently with the DBCSR library?
If yes, you can re-use the already present "infrastructure" in `exts/`, but this would require that you adapt a somewhat compatible build-system. If you instead have to run either `CMake` or `autoconf` and the library is still optional, the "SIRIUS"-approach might be better suited.

Since this started as an in-house code, please make sure that the library has a GPL-2 compatible license.

**Comment by oschuett** [comment]
I strongly advice against the submodule approach because it only combines the disadvantage. You get the constraints (build system) but not the daily tests. Hence, for DBCSR we [recently decided](https://github.com/cp2k/cp2k/issues/351) to make it a fully independent library.

**Comment by dev-zero** [comment]
> ... but not the daily tests.

can you please elaborate on this?

**Comment by oschuett** [comment]
Changes to DBCSR are not automatically included in our daily dashboard tests. It requires a separate commit to CP2K to point the submodule to a new version. Since this is a manual process it happens only occasionally.

Take e.g. the recent https://github.com/cp2k/cp2k/commit/2824b54caf63b9adad79685dd3db4f417b4ac058 which changed the submodule from https://github.com/cp2k/dbcsr/commit/e4db53ef9f983fcfda72edc747d598935038b77d to https://github.com/cp2k/dbcsr/commit/98083b415e538cd8f38592fe02203572fad8588c. In between those two versions are 59 commits from the last 4 weeks. Through the dashboard we discovered that sometime within those 4 weeks a bug was introduced to DBCSR. In contrast, we would have found this bug immediately had DBCSR been in the CP2K main repository.

Quick feedback from extensive tests is a big productivity advantage.

**Comment by dev-zero** [comment]
> Changes to DBCSR are not automatically included in our daily dashboard tests. It requires a separate commit to CP2K to point the submodule to a new version. Since this is a manual process it happens only occasionally.

The same applies for other external libraries like SIRIUS and will also apply to DBCSR once it's really an external library.

But for each submodule you have a tracking branch specified, so you could setup a tester doing `git submodule update --remote` to test against the HEAD of the respective branch instead.

**Comment by cschran** [comment]
Thank you for your feedback. We agree that extensive testing via CI and dashboard is a clear advantage of a direct implementation. The downside for us is that we would need to manually maintain synchronization with our fitting code.

We will look into the direction to keep our implementation exclusively in cp2k and try to separate the core routines and data structure from its cp2k plumbing. That will hopefully allow us to maintain the overlap with our in-house project and make the synchronization easier.

**Comment by oschuett** [comment]
> so you could setup a tester ... against the HEAD of the respective branch instead.

I don't think this would work. Who would maintain this? The 3rd party project doesn't know or care about our tests. Automatic tests require a certainly level of standardization. In our case this means using our repository, build system, and coding conventions.

> The same applies for other external libraries like SIRIUS and will also apply to DBCSR once it's really an external library.

Yes, these projects are relying on their unit tests. As we have [recently seen](https://github.com/cp2k/cp2k/commit/2824b54caf63b9adad79685dd3db4f417b4ac058), it's difficult to get good coverage with unit tests alone. However, the real challenge with testing is heterogeneity, which in the foreseeable future will only go up.

**Comment by cschran** [comment]
After quite some time, we've finally managed to get the implementation into a stage where it can be shared with others. As discussed roughly a year ago, it's self-contained and fully functional within cp2k.

Since the commit is on the larger side, we would appreciate any feedback. cp2k/cp2k#1161

Would be great, if this could still make it into the next release :+1: 

Thanks and all the best,
Christoph
